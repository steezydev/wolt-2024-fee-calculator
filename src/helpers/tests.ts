const localStorageMock = (function () {
    let store: Record<string, string> = {};
    return {
        getItem(key: string) {
            return store[key] || null;
        },
        setItem(key: string, value: string) {
            store[key] = value.toString();
        },
        clear() {
            store = {};
        },
    };
})();

const matchMediaMock = jest.fn().mockImplementation((query) => ({
    matches: query === '(prefers-color-scheme: dark)',
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
}));

export const mockContext = () => {
    beforeAll(() => {
        Object.defineProperty(window, 'localStorage', {
            value: {
                getItem: jest.fn().mockImplementation((key: string) => {
                    return localStorageMock.getItem(key);
                }),
                setItem: jest
                    .fn()
                    .mockImplementation((key: string, value: string) => {
                        localStorageMock.setItem(key, value);
                    }),
                clear: localStorageMock.clear,
            },
            writable: true,
        });

        window.matchMedia = matchMediaMock;
    });

    beforeEach(() => {
        (window.localStorage.getItem as jest.Mock).mockClear();
        (window.localStorage.setItem as jest.Mock).mockClear();
        matchMediaMock.mockClear();
    });
};

export const setupFakeDate = () => {
    beforeEach(() => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2024-01-01'));
    });

    afterEach(() => {
        jest.useRealTimers();
    });
};

export const mockScrollTo = () => {
    beforeEach(() => {
        Element.prototype.scrollTo = jest.fn();
    });

    afterEach(() => {
        (Element.prototype.scrollTo as jest.Mock).mockRestore();
    });
};
