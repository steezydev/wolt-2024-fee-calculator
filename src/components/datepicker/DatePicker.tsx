import Calendar from '@/components/calendar/Calendar';
import DateField from '@/components/datefield/DateField';
import Tabs from '@/components/tabs/Tabs';
import TimeSelector from '@/components/timeselector/TimeSelector';
import B3 from '@/components/typography/B3';
import { classNames } from '@/helpers/classNames';
import useClickOutside from '@/hooks/useClickOutside';
import { DatePickerProps } from '@/types/props/DatePickerProps';
import { RenderInputProps } from '@/types/props/RenderInputProps';
import React, { useState } from 'react';

const DatePicker = ({
  value,
  renderInput,
  onDateChange,
  onTimeChange,
  onChange,
  className,
}: DatePickerProps) => {
  const [isPickerOpen, setIsPickerOpen] = useState(false);
  const [isFocused, setIsFocused] = useState(false);

  const ref = useClickOutside<HTMLDivElement>(() => setIsPickerOpen(false));

  const handleBlur = () => {
    setIsFocused(false);
    if (!ref.current?.focus) {
      setIsPickerOpen(false);
    }
  };

  const handleFocus = () => {
    setIsPickerOpen(true);
    setIsFocused(true);
  };

  // Add onBlur and onFocus handlers to renderInput on top of existing ones
  const additiveRenderInput = (props: RenderInputProps) => {
    const onBlur = props.onBlur;
    const onFocus = props.onFocus;
    return renderInput({
      ...props,
      onBlur: (e: React.FocusEvent<HTMLInputElement>) => {
        handleBlur();
        onBlur && onBlur(e);
      },
      onFocus: (e: React.FocusEvent<HTMLInputElement>) => {
        handleFocus();
        onFocus && onFocus(e);
      },
    });
  };

  return (
    <div
      data-invalid={false}
      data-focused={isFocused}
      className={classNames('w-full relative', className)}
    >
      <DateField
        date={value}
        onDateChange={onChange}
        renderInput={additiveRenderInput}
      />
      {isPickerOpen && (
        <div
          className='absolute bg-white dark:bg-black border-2 rounded-xl border-primary-300 top-16 flex flex-row h-[360px] w-80 p-2'
          ref={ref}
        >
          <Tabs className='w-full'>
            <Tabs.Tab
              id='datePickerTabButton'
              ariaLabel='Open date picker tab'
              value='datePickerTabButton'
            >
              <B3>Date</B3>
            </Tabs.Tab>
            <Tabs.Panel value='datePickerTabButton'>
              <Calendar
                weekStartsOn={1}
                showOutsideDays
                selectedDate={value}
                onChange={onDateChange}
              />
            </Tabs.Panel>
            <Tabs.Tab
              id='timePickerTabButton'
              ariaLabel='Open time picker tab'
              value='timePickerTabButton'
            >
              <B3>Time</B3>
            </Tabs.Tab>
            <Tabs.Panel value='timePickerTabButton' className='p-2'>
              <TimeSelector className='' time={value} onChange={onTimeChange} />
            </Tabs.Panel>
          </Tabs>
        </div>
      )}
    </div>
  );
};

export default DatePicker;
