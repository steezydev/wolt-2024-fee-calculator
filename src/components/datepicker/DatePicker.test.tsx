import DatePicker from './DatePicker';
import { dateFieldRenderTests } from './DatePicker.tests';

dateFieldRenderTests('DatePicker', DatePicker);
