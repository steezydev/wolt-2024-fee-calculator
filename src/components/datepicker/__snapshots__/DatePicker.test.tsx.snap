// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker when asked to render renders 1`] = `
<div
  className="w-full relative"
  data-focused={false}
  data-invalid={false}
>
  <div
    className="flex gap-2"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="17.01.2024 12:00"
    />
    <button
      aria-label="Open datepicker modal"
      className="bg-primary-100 text-lg font-bold py-3 rounded-xl hover:bg-primary-300/20 active:bg-primary-300/50 disabled:bg-neutral-light-100 disabled:dark:bg-neutral-dark-100 disabled:text-white/40 disabled:cursor-not-allowed focus-visible:ring-4 focus-visible:ring-focus focus:outline-none focus-visible:outline-none aspect-square flex justify-center items-center px-3 h-full flex-shrink-0 flex-grow text-primary-300"
      data-test-id="openDatePickerButton"
      disabled={false}
      id="openDatePickerButton"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="w-6 h-6 text-inherit"
        fill="none"
        height="24"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 2v4"
        />
        <path
          d="M16 2v4"
        />
        <rect
          height="18"
          rx="2"
          width="18"
          x="3"
          y="4"
        />
        <path
          d="M3 10h18"
        />
        <path
          d="M10 16h4"
        />
        <path
          d="M12 14v4"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`DatePicker when asked to render with additional class name renders with additional class name 1`] = `
<div
  className="w-full relative some-class"
  data-focused={false}
  data-invalid={false}
>
  <div
    className="flex gap-2"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="17.01.2024 12:00"
    />
    <button
      aria-label="Open datepicker modal"
      className="bg-primary-100 text-lg font-bold py-3 rounded-xl hover:bg-primary-300/20 active:bg-primary-300/50 disabled:bg-neutral-light-100 disabled:dark:bg-neutral-dark-100 disabled:text-white/40 disabled:cursor-not-allowed focus-visible:ring-4 focus-visible:ring-focus focus:outline-none focus-visible:outline-none aspect-square flex justify-center items-center px-3 h-full flex-shrink-0 flex-grow text-primary-300"
      data-test-id="openDatePickerButton"
      disabled={false}
      id="openDatePickerButton"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="w-6 h-6 text-inherit"
        fill="none"
        height="24"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 2v4"
        />
        <path
          d="M16 2v4"
        />
        <rect
          height="18"
          rx="2"
          width="18"
          x="3"
          y="4"
        />
        <path
          d="M3 10h18"
        />
        <path
          d="M10 16h4"
        />
        <path
          d="M12 14v4"
        />
      </svg>
    </button>
  </div>
</div>
`;
